design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,495 @@
+
+Without getting too deep into specifics, early rendering passes allow the developer to use the power of the rasterizer for their own data.  Later rendering passes are typically using to decorate the map with effects.
+
+#### The way it is now:
+The [low level rendering logic](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/renderer_impl.cpp) in the toolkit seems to only support one explicit rendering pass.  
+
+However, there is the [RenderPass](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/gl/render_pass.hpp) so some notion of this exists in the toolkit, but it's not clear how complete that is.  It may depend on ordering to work things out, rather than the explicit command buffer filling and fences we would use in a more modern approach.
+
+In any case, we need this filled out more explicitly for the render targets.
+
+#### Required changes:
+To flexibly implement multiple rendering passes we will need to:
+- Keep track of them by name and order
+- Implement Render Targets (see below)
+- Outputs from one pass as inputs to another
+- Optimize out passes without geometry
+- Expose all of this outside the toolkit
+
+#### Benefits
+Similar to what was discussed above:
+- Using the rasterizer to sort out your data ahead of time.  We do this a lot with weather data.
+- Reducing the size of a very costly shader.  The bigger your shader, the longer it takes.  If you have some fancy effect, it can probably be rendered at lower resolution and then laid over the top of the final scene.
+- Doing post-process effects on the map.  In games you might think of lens warping, wipes, and such.  For mapping, this tends to be more merging data layers or things like glow.
+
+### Off Screen Render Targets
+
+_Addresses core functionality [#5](#core)._
+
+The toolkit already has support for one offscreen render target.  You can either set things up “normally” where the toolkit draws to a window or you can set things up to draw to memory and it will hand you back the resulting image.  This is a little like that, but in real time.
+
+Having an explicit off screen render target lets the developer do something like a snapshot in real time.  That is, run some sort of logic on the data being rendered before it gets its final visual representation.
+
+#### Required Changes:
+To support Rendering Passes we’ll need Render Targets we can access in a way similar to a texture.  They’ll be set up with:
+- Settings such as bit depth and width/height (you don’t always want full screen resolution)
+- Allowing geometry to specify which one(s) they’ll render to as a “target”
+- Query and snapshot
+
+This may seem a bit of a detour, but it’s fairly easy to get the basics of it in place and allow for further development as developers might add new shaders and varieties of targets.
+
+#### Benefits
+It's a core idea in many rendering toolkits, but for MapLibre Native it will let us:
```

Comment:
Would be helpful for me if you could define somewhere at the beginning of this document how you use the words "toolkit" and "SDK".

Reply:
Added something at the beginning.